cmd/commands/types.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/inngest/inngest/pkg/function"
+	"github.com/spf13/cobra"
+)
+
+func NewCmdTypes() *cobra.Command {
+	root := &cobra.Command{
+		Use:     "types",
+		Short:   "Generate types from your Inngest Cloud account.",
+		Example: "inngest types typescript",
+	}
+
+	typescript := &cobra.Command{
+		Use:     "typescript",
+		Aliases: []string{"ts"},
+		Short:   "Generate TypeScript types in a .ts file",
+		Run:     doTypescript,
+	}
+
+	root.AddCommand(typescript)
+
+	root.PersistentFlags().StringP("output", "o", "./__generated__/inngest.ts", "Specify the location of the generated .ts file")
+
+	return root
+}
+
+// Write the given `types` to the given `absPath`. Will try to create any
+// required directories.
+func writeTypes(types string, absPath string) error {
+	// Try to create the folder and file with these TS types
+	dirRequired := filepath.Dir(absPath)
+
+	err := os.MkdirAll(dirRequired, 0755)
+	if err != nil {
+		return fmt.Errorf("couldn't create directory for output file; %w", err)
+	}
+
+	err = os.WriteFile(absPath, []byte(types), 0755)
+	if err != nil {
+		return fmt.Errorf("couldn't create output file; %w", err)
+	}
+
+	fmt.Println("Successfully created types at", absPath)
+
+	return nil
+}
+
+// Checks that the given `relPath` is valid and contains the valid suffix
+// depending on the types being generated. For example, it should be invalid to
+// try to create TypeScript types as a `.go` file.
+//
+// `suffix` should be the file extension without the `.`, e.g. `ts` or `go`.
+func checkOutputTarget(relPath string, suffix string) (string, error) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("failed to get working directory; %w", err)
+	}
+
+	absPath := filepath.Join(cwd, relPath)
+	if !strings.HasSuffix(absPath, suffix) {
+		return "", fmt.Errorf("output file must be a .ts file")
+	}
+
+	return absPath, nil
+}
+
+func doTypescript(cmd *cobra.Command, args []string) {
+	fmt.Println(cli.EnvString())
+
+	absPath, err := checkOutputTarget(cmd.Flag("output").Value.String(), "ts")
+	if err != nil {
+		log.Fatalln("\n" + cli.RenderError(err.Error()) + "\n")
+	}
+
+	types, err := typescript(cmd, args)
+	if err != nil {
+		log.Fatalln("\n" + cli.RenderError(err.Error()) + "\n")
+	}
+
+	err = writeTypes(types, absPath)
+	if err != nil {
+		log.Fatalln("\n" + cli.RenderError(err.Error()) + "\n")
+	}
+}
+
+// Fetch events and generate TypeScript types.
+//
+// TODO Extract fetching events here for easier reuse when adding other
+// languages.
+func typescript(cmd *cobra.Command, args []string) (string, error) {
+	ctx := cmd.Context()
+
+	// Try fetching events
+	events, err := initialize.FetchEvents()
+	if err != nil {
+		if !strings.Contains(err.Error(), "not logged in") {
+			return "", err
+		}
+
+		// If we got an error because we're not logged in, try logging in
+		// and attempt once more
+		DeviceAuth(ctx)
+		events, err = initialize.FetchEvents()
+		if err != nil {
+			return "", err
```

Comment:
Hmm, what if you only want to get public events here?  Should we force a login? 🤔